ui.apps/src/main/content/jcr_root/apps/venia/components/structure/navigation/v1/navigation/clientlibs/js/navigation.js

```diff
@@ -0,0 +1,110 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+
+(function () {
+    "use strict";
+    const selectors = {
+        navigationTrigger: '.header__primaryActions .navTrigger__root',
+        navigationRoot: 'aside.navigation__root',
+        shadowTreeRoot: '.categoryTree__root--shadow',
+        categoryTreeRoot: '.categoryTree__root',
+        activeNavigation: '.categoryTree__root .categoryTree__tree',
+        shadowNavigations: '.categoryTree__root--shadow .categoryTree__tree',
+        backNavigationButton: '.navigation__header .trigger__root--back',
+        backNavigationEmpty: '.navigation__header .trigger__root--back--empty',
+        closeNavigationButton: '.navigation__header .trigger__root--close',
+        downNavigationButton: '.categoryLeaf__root button'
+    };
+
+    const CSS_NAVIGATION_OPEN = 'navigation__root_open';
+
+    let backNavigationButton;
+    let backNavigationEmpty;
+    let categoryTreeRoot;
+    let shadowTreeRoot;
+    let navigationPanel;
+
+    function setNavigationPanelVisible(visible) {
+        if (visible) {
+            navigationPanel.classList.add(CSS_NAVIGATION_OPEN);
+        } else {
+            navigationPanel.classList.remove(CSS_NAVIGATION_OPEN);
+        }
+    }
+
+    function setVisible(element, visible) {
+        if (visible) {
+            element.style.display = 'block';
+        } else {
+            element.style.display = 'none';
+        }
+    }
+
+    function getActiveNavigation() {
+        return document.querySelector(selectors.activeNavigation);
+    }
+
+    function configureBackNavigation() {
+        let hasParent = getActiveNavigation().dataset.parent;
+        if (hasParent) {
+            setVisible(backNavigationButton, true);
+            setVisible(backNavigationEmpty, false);
+        } else {
+            setVisible(backNavigationButton, false);
+            setVisible(backNavigationEmpty, true);
+        }
+    }
+
+    function activateNavigation(id) {
+        let navigation = document.querySelector(selectors.shadowNavigations + '[data-id="' + id + '"]');
+        if (navigation) {
+            shadowTreeRoot.append(getActiveNavigation());
+            categoryTreeRoot.append(navigation);
+
+            configureBackNavigation();
+        }
+    }
+
+    function backNavigation() {
+        let id = getActiveNavigation().dataset.parent;
+        activateNavigation(id);
+    }
+
+    function downNavigation(event) {
+        const id = event.target.parentElement.parentElement.dataset.id;
+        activateNavigation(id);
+    }
+
+    function onDocumentReady() {
```

Comment:
Is there a reason why you decided against encapsulating the client-side behavior of this component in a class or prototype like in the other components?

Reply:
I can add that, since we seem to generally prefer a class.

Reply:
Having a class makes it easier to write tests this when we'll introduce unit testing.